File: app/src/main/java/com/chopcode/rutago/app/viewmodels/passenger/EditProfileViewModel.java
package com.chopcode.rutago.app.viewmodels.passenger;

import android.util.Log;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.chopcode.rutago.app.config.MyApp;
import com.chopcode.rutago.app.models.User;
import com.chopcode.rutago.app.services.user.UserService;
import com.chopcode.rutago.app.R;

import java.util.HashMap;
import java.util.Map;

/**
 * Edit Profile ViewModel (Passenger)
 *
 * Gestor del flujo de actualización de información básica del pasajero.
 * Responsabilidades:
 * - Recuperar el estado actual del perfil para pre-cargar el formulario.
 * - Validar reglas de negocio sobre campos obligatorios (Nombre y Teléfono).
 * - Delegar al UserService la persistencia de cambios en el nodo /usuarios/.
 * - Proporcionar feedback reactivo sobre el éxito o fallo de la operación de edición.
 */
public class EditProfileViewModel extends ViewModel {
    private static final String TAG = "EditProfileVM";

    /** Datos actuales del usuario para inicializar la vista. */
    private final MutableLiveData<User> userData = new MutableLiveData<>();
    
    private final MutableLiveData<Boolean> isLoading = new MutableLiveData<>(false);
    
    /** Notifica el éxito de la transacción para cerrar la pantalla de edición. */
    private final MutableLiveData<Boolean> updateSuccess = new MutableLiveData<>(false);
    
    private final MutableLiveData<String> error = new MutableLiveData<>();

    private final UserService userService;

    public EditProfileViewModel() {
        this.userService = new UserService();
    }

    public LiveData<User> getUserData() { return userData; }
    public LiveData<Boolean> getIsLoading() { return isLoading; }
    public LiveData<Boolean> getUpdateSuccess() { return updateSuccess; }
    public LiveData<String> getError() { return error; }

    /**
     * Recupera la información de perfil del pasajero logueado.
     */
    public void loadUserProfile() {
        String userId = MyApp.getCurrentUserId();
        if (userId == null) {
            error.setValue(MyApp.getAppContext().getString(R.string.usuario_no_autenticado));
            return;
        }

        isLoading.setValue(true);
        userService.loadUserData(userId, new UserService.UserDataCallback() {
            @Override
            public void onUserDataLoaded(User usuario) {
                userData.postValue(usuario);
                isLoading.postValue(false);
            }

            @Override
            public void onError(String errorMsg) {
                Log.e(TAG, "❌ Error al cargar perfil: " + errorMsg);
                error.postValue(MyApp.getAppContext().getString(R.string.error_prefijo, errorMsg));
                isLoading.postValue(false);
            }
        });
    }

    /**
     * Ejecuta la validación y persistencia del nuevo perfil.
     * @param nuevoNombre Nombre actualizado del pasajero.
     * @param nuevoTelefono Número de contacto actualizado.
     */
    public void updateProfile(String nuevoNombre, String nuevoTelefono) {
        String userId = MyApp.getCurrentUserId();
        if (userId == null) {
            error.setValue(MyApp.getAppContext().getString(R.string.usuario_no_autenticado));
            return;
        }

        // Validaciones de integridad de datos
        if (nuevoNombre == null || nuevoNombre.trim().isEmpty()) {
            error.setValue(MyApp.getAppContext().getString(R.string.nombre_obligatorio));
            return;
        }

        if (nuevoTelefono == null || nuevoTelefono.trim().isEmpty()) {
            error.setValue(MyApp.getAppContext().getString(R.string.telefono_obligatorio));
            return;
        }

        isLoading.setValue(true);
        Log.d(TAG, "🔄 Persistiendo cambios de perfil para: " + userId);

        userService.updateUserProfile(userId, nuevoNombre, nuevoTelefono, new UserService.UserUpdateCallback() {
            @Override
            public void onSuccess() {
                updateSuccess.postValue(true);
                isLoading.postValue(false);
            }

            @Override
            public void onError(String errorMsg) {
                Log.e(TAG, "❌ Fallo en la actualización: " + errorMsg);
                error.postValue(MyApp.getAppContext().getString(R.string.error_actualizar, errorMsg));
                isLoading.postValue(false);
            }
        });
    }
}

File: app/src/main/java/com/chopcode/rutago/app/ui/viewmodels/passenger/EditProfileViewModel.kt
package com.chopcode.rutago.app.ui.viewmodels.passenger

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.chopcode.rutago.app.R
import com.chopcode.rutago.app.config.MyApp
import com.chopcode.rutago.app.models.User
import com.chopcode.rutago.app.services.user.UserService
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/**
 * 🧠 VIEWMODEL: EditProfileViewModel
 * Gestor del flujo de actualización de información básica del pasajero.
 */
class EditProfileViewModel : ViewModel() {

    private val _uiState = MutableStateFlow(EditProfileUiState())
    val uiState: StateFlow<EditProfileUiState> = _uiState.asStateFlow()

    private val userService = UserService()

    init {
        loadUserProfile()
    }

    fun loadUserProfile() {
        val userId = MyApp.getCurrentUserId()
        if (userId == null) {
            _uiState.update { it.copy(error = "Usuario no autenticado") }
            return
        }

        _uiState.update { it.copy(isLoading = true) }
        
        userService.loadUserData(userId, object : UserService.UserDataCallback {
            override fun onUserDataLoaded(user: User?) {
                user?.let { u ->
                    _uiState.update { 
                        it.copy(
                            isLoading = false,
                            currentName = u.nombre ?: "N/A",
                            currentPhone = u.telefono ?: "N/A",
                            currentEmail = u.email ?: "N/A",
                            name = u.nombre ?: "",
                            phone = u.telefono ?: "",
                            email = u.email ?: ""
                        )
                    }
                }
            }

            override fun onError(errorMsg: String?) {
                _uiState.update { it.copy(isLoading = false, error = errorMsg) }
            }
        })
    }

    fun onNameChange(newName: String) {
        _uiState.update { it.copy(name = newName) }
    }

    fun onPhoneChange(newPhone: String) {
        _uiState.update { it.copy(phone = newPhone) }
    }

    fun updateProfile() {
        val userId = MyApp.getCurrentUserId() ?: return
        val state = _uiState.value
        
        if (!state.isFormValid) return

        _uiState.update { it.copy(isLoading = true) }
        
        userService.updateUserProfile(userId, state.name, state.phone, object : UserService.UserUpdateCallback {
            override fun onSuccess() {
                _uiState.update { it.copy(isLoading = false, updateSuccess = true) }
            }

            override fun onError(errorMsg: String?) {
                _uiState.update { it.copy(isLoading = false, error = errorMsg) }
            }
        })
    }
}
